### ISSUE TITLE
Package architecture naming convention

### ISSUE THREAD

**Issue by igagis** [Alexpux/MINGW-packages#4787]
This is a question to clarify the package naming convention for different architectures.

I'm confused about the thing that in http://repo.msys2.org/mingw/x86_64/ repo all the packages have architecture `any`, while x86_64 is a part of the package name. Why is that so?

In http://repo.msys2.org/msys/x86_64/ repo packages are named correctly with x86_64 architecutre.

Could you explain in more details why mingw packages are named like that and have architecture `any`?

I'm confused about how to name my own packages...

**Comment by mati865** [comment]
MSYS2 can be either i686 or x86_64 and mingw packages are independent of MSYS2 arch. `any` simplifies packaging and avoids duplicated work.

Consider following examples:
* `arch=('any')` - you create just one `mingw-w64-i686-name-version-any.pkg.tar.xz` and one `mingw-w64-x86_64-name-version-any.pkg.tar.xz` package. They can be installed on i686 or x86_64 MSYS2 without anything special.
* `arch=('i686', 'x86_64')` - you'd have to create 2 packages for each arch: `mingw-w64-i686-name-version-i686.pkg.tar.xz` and `mingw-w64-x86_64-name-version-i686.pkg.tar.xz` for i686 MSYS2 (they won't install on x86_64 MSYS2) and `mingw-w64-i686-name-version-x86_64.pkg.tar.xz` and `mingw-w64-i686-x86_64-version-x86_64.pkg.tar.xz` for x86_64 MSYS2 (they won't install on i686 MSYS2).

**Comment by igagis** [comment]
It all seems clear, except one thing:

examining files from `mingw-w64-x86_64-gcc` package I found that it contains executables, e.g. `/mingw64/bin/gcc.exe`. So, since the package has architecture `any` it means that it can be installed even on 32 bit machine in msys. But how to run the executable then if it is built for 64bit machine?

**Comment by StarWolf3000** [comment]
They're PE32+ executables, means they are 64-bit binaries and can only be run on a 64-bit Windows. But everything you build with them can be targeted both i686 and x86-64 (arch=any).
mingw-w64-i686-* are 32-bit binaries. They can target both i686 and x86-x64 (arch=any).

**Comment by StarWolf3000** [comment]
MSYS2:
`<name>-<ver>-i686.pkg.tar.xz` = `<name>` with version `<ver>` to be run on architecture `i686` (32-bit).
Compilation of MSYS2 packages for both i686 and x86-64 are supported by this GCC.

`<name>-<ver>-x86_64.pkg.tar.xz` = `<name>` with version `<ver>` to be run on architecture `x86-64` (64-bit).
Compilation of MSYS2 packages for both i686 and x86-64 are supported by this GCC.

----------------------------------------------

MINGW:
`mingw-w64-i686-<name>-<ver>-any.pkg.tar.xz` = `<name>` with version `<ver>` to be run on architecture `i686` (32-bit).
Compilation of MINGW packages for both i686 and x86-64 are supported by this GCC.

`mingw-w64-x86_64-<name>-<ver>-any.pkg.tar.xz` = `<name>` with version `<ver>` to be run on architecture `x86-64` (64-bit).
Compilation of MINGW packages for both i686 and x86-64 are supported by this GCC.

----------------------------------------------

You can build 32-bit and 64-bit files with both the i686 and x86_64 environments, but you can only install and run the x86_64 environment on a 64-bit Windows (yes, you can build 64-bit applications on a 32-bit Windows with this).

**Comment by mati865** [comment]
@igagis no, `arch` in `<mingw_arch>-<name>-<ver>-<arch>.pkg.tar.xz` means MSYS2 arch. You can run 32 bit MSYS2 on 64 bit system and use 64 bit mingw pacakges.

`mingw_arch` would refer to host and target architecture, 64 bit system (with 32 or 64 bit MSYS2) can run any package and 32 bit system (with 32 bit MSYS2) can only run i686 packages.

Think of it like distribution inside distribution. MSYS2 packages are build for specific MSYS2 arch and you cannot install `<mingw_arch>-<name>-<ver>-i686.pkg.tar.xz` in x86_64 MSYS2 or the other way round.
Now `/mingw32` and `/mingw64` "distributions" are independent of the main distribution (MSYS2) hence they don't rely on `arch` but have their own `mingw_arch` to distinguish them.

> And my understanding also that mingw32 toolchain builds 32 bit binaries and mingw64 toolchain builds 64 bit binaries. Is that right?

That's right.

---
> You can build 32-bit and 64-bit files with both the i686 and x86_64 environments, but you can only install and run the x86_64 environment on a 64-bit Windows (yes, you can build 64-bit applications on a 32-bit Windows with this).

Technically it's possible but requires a lot of manual tinkering because you need all tools to support cross compilation and configure `/mingw32` sysroot for them.

**Comment by igagis** [comment]
ok, so `mingw64` cannot be run on 32 bit machine, right?
